refactor(ivy): combine contentQueries and contentQueriesRefresh functions

AndrewKushnir opened the pull request:
Prior to this update we had separate contentQueries and contentQueriesRefresh functions to handle creation and update phases. This approach was inconsistent with View Queries, Host Bindings and Template functions that we generate for Component/Directive defs. Now the mentioned 2 functions are combines into one (contentQueries), creation and update logic is separated with RenderFlags (similar to what we have in other generated functions).

This PR addresses FW-923.

There are few other steps I'd like to take in future PRs as further refactoring:

- rename `viewQuery` -> `viewQueries` (for consistency)
- at runtime, we use `QUERIES` const to refer to View Queries and `CONTENT_QUERIES` const for Content Queries - we should probably use `VIEW_QUERIES` and `CONTENT_QUERIES` to have a clear separation
- in `meta` fields in compiler we use `queries` to refer to Content Queries and `viewQueries` to refer to View Queries - we can unify that too
- the logic to generate View and Content Queries in Compiler is very similar now (with a few exceptions like the names of functions we use inside query function, like `viewQuery` vs `contentQuery` and different set of params that the function accepts)
- (may be) get rid of `directiveIndex` argument in `contentQueries` function to further unify `viewQueries` and `contentQueries`.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
refactor(ivy): combine contentQueries and contentQueriesRefresh functions

Prior to this update we had separate contentQueries and contentQueriesRefresh functions to handle creation and update phases. This approach was inconsistent with View Queries, Host Bindings and Template functions that we generate for Component/Directive defs. Now the mentioned 2 functions are combines into one (contentQueries), creation and update logic is separated with RenderFlags (similar to what we have in other generated functions).

AndrewKushnir committed:
fixup! refactor(ivy): combine contentQueries and contentQueriesRefresh functions

kara commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -131,12 +137,11 @@ export interface DirectiveDef<T> extends BaseDef<T> {
   /**
    * Function to create instances of content queries associated with a given directive.
```
Nit: Update docs comment?

```suggestion
   * Function to create and refresh content queries associated with a given directive.
```

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -134,13 +134,15 @@ export function setHostBindings(tView: TView, viewData: LView): void {
-      directiveDef.contentQueriesRefresh !(directiveDefIdx - HEADER_OFFSET);
+      const directiveIndex = directiveDefIdx - HEADER_OFFSET;
+      directiveDef.contentQueries !(
+          RenderFlags.Update, loadInternal(lView, directiveIndex), directiveDefIdx);
```
`loadInternal` just adds the `HEADER_OFFSET` back. Since we already have a raw index, seems like we should use it directly (and remove extra var).

```suggestion
          RenderFlags.Update, lView[directiveDefIndex], directiveDefIdx);
```

kara reviewed (changes requested):
2 small things, otherwise looks good

AndrewKushnir committed:
fixup! refactor(ivy): combine contentQueries and contentQueriesRefresh functions

AndrewKushnir committed:
fixup! refactor(ivy): combine contentQueries and contentQueriesRefresh functions

AndrewKushnir commented:
@kara thanks for the review. I've addressed your comments and rebased this PR on top of the latest master. Could you please take another look (esp. `instructions.ts`, since I had to resolve some conflicts there)? Thank you.

kara reviewed (approved):
LGTM

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -64,9 +64,10 @@ function baseDirectiveFields(
-  definitionMap.set('contentQueries', createContentQueriesFunction(meta, constantPool));
-
-  definitionMap.set('contentQueriesRefresh', createContentQueriesRefreshFunction(meta));
+  if (meta.queries.length) {
```
```suggestion
  if (meta.queries.length > 0) {
```

I prefer to be explicit in conditionals, for 3 reasons:

1. Readability - it's obvious looking at the explicit comparison exactly what condition is being checked. Whereas looking at `if (foo)`, it's not really clear which values `foo` can take and which of them are being 

2. Correctness - it's easy to make mistakes and write expressions like `if (!template)`, which fails if `template` is an empty string.

3. Performance - JavaScript VMs emit far more code to support `if (x)` than `if (x > 0)`, because of the range of possible values. There's not a big difference, but it does exist.

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -510,56 +511,39 @@ function convertAttributesToExpressions(attributes: any): o.Expression[] {
-
-      statements.push(refreshQueryAndUpdateDirective.toStmt());
-    });
+  meta.queries.forEach((query: R3QueryMetadata) => {
```
I've been switching more compiler code to use `for (const query of meta.queries)` instead of `.forEach`.

AndrewKushnir committed:
fixup! refactor(ivy): combine contentQueries and contentQueriesRefresh functions

AndrewKushnir commented:
@alxhub thanks for the review. I've updated the code based on your feedback. Thank you.

AndrewKushnir commented:
[Presubmit](http://test/OCL:233687463:BASE:233687568:1550036505100:8247030)

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -80,7 +80,7 @@ export function refreshDescendantViews(lView: LView) {
 
     // Content query results must be refreshed before content hooks are called.
-    refreshContentQueries(tView);
+    refreshContentQueries(tView, lView);
```
Now we could just pass `lView` as `tView` can be retrieved from `lView`

AndrewKushnir replied:
I kept `tView` for now, since other functions around this code have the same pattern. I think we can refactor that altogether with other fns.

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -134,13 +134,13 @@ export function setHostBindings(tView: TView, viewData: LView): void {
       const directiveDefIdx = tView.contentQueries[i];
       const directiveDef = tView.data[directiveDefIdx] as DirectiveDef<any>;
-      directiveDef.contentQueriesRefresh !(directiveDefIdx - HEADER_OFFSET);
+      directiveDef.contentQueries !(RenderFlags.Update, lView[directiveDefIdx], directiveDefIdx);
```
nit: add assert for `directiveDef.contentQueries`?

AndrewKushnir replied:
Thanks Pawel, I've added assert to make sure `directiveDef.contentQueries` fn is defined.

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -534,17 +534,17 @@ export function elementContainerStart(
     lView[QUERIES] = currentQueries.clone();
   }
-  executeContentQueries(tView, tNode);
+  executeContentQueries(tView, tNode, lView);
```
Drop `tView` from arguments?

AndrewKushnir replied:
I kept `tView` for now, since other functions around this code have the same pattern. I think we can refactor that altogether with other fns.

pkozlowski-opensource reviewed (approved):
Left a couple of nit-comments but LGTM overall

AndrewKushnir committed:
fixup! refactor(ivy): combine contentQueries and contentQueriesRefresh functions

AndrewKushnir committed:
fixup! refactor(ivy): combine contentQueries and contentQueriesRefresh functions
